### PR TITLE
Show Private Conversations

### PR DESCRIPTION
**Pull request by borisyankov**
* Prefetch private conversations
* User Search in new Screen
* Add Group chat item display

### CODE REVIEW

**Comment by smarx** [comment]
*Automated message from Dropbox CLA bot*

@borisyankov, it looks like you've already signed the Dropbox CLA. Thanks!

File: src/chat/chatSelectors.js [hunk 1]
```diff
@@ -13,3 +16,16 @@ export const getPointer = (state) => {
 
   return [messages[0].id, messages[messages.length - 1].id];
 };
+
+export const getRecentConversations = (state) => {
+  const selfEmail = state.account[0].email;
+  const messages = state.chat.messages[JSON.stringify(specialNarrow('private'))] || [];
+  const recipientSet = messages.reduce((accumulator, x) => {
```
Comment (neerajwahi) [inline src/chat/chatSelectors.js]: This is hard to read as a reduce pattern. Reduce usually implies that N elements get condensed to 1 element.

I think it's much easier to read if made imperative:
```
const recipients = new Set();
for (const message of messages) {
  ...
  recipients.add(normalized);
}
return recipients;
```

Alternatively:
```
return unique(
  messages
  .map((msg) => normalizeRecipients(msg.display_recipient))
  .filter((r) => r.email !== selfEmail)
);
```
where
```
unique = (list) => [...new Set(list)];
```

File: src/actions.js [hunk 2]
```diff
@@ -3,4 +3,4 @@ export * from './account/accountActions';
 export * from './events/eventActions';
 export * from './nav/navActions';
 export * from './message-list/messagesActions';
-export * from './userlist/userListActions';
+export * from './users/userListActions';
```
Comment (neerajwahi) [inline src/actions.js]: I like the rename to users (from userlist) 👍

File: src/conversations/ConversationList.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    const dataSource = ds.cloneWithRows(conversations);
+
+    return (
+      <ListView
+        enableEmptySections
+        style={styles.container}
+        dataSource={dataSource}
+        pageSize={12}
+        renderRow={(email => {
+          const renderFunc = (email.indexOf(',') === -1) ?
```
Comment (neerajwahi) [inline src/conversations/ConversationList.js]: Can you add a comment here? Something like "If there are multiple emails, render a group". Future readers may not understand why we're looking for `,` in `email`

File: src/conversations/ConversationList.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+export default class ConversationList extends Component {
+
+  props: {
+    conversations: string[],
+    realm: string,
+    users: any[],
+    onNarrow: (email: string) => void,
+  }
+
+  renderUserItem = ({ email, users, realm, narrow, onNarrow }) => {
```
Comment (neerajwahi) [inline src/conversations/ConversationList.js]: `renderUserItem` and `renderUserGroup` seem better suited as separate (stateless) components

File: src/conversations/ConversationsCard.js [hunk 5]
```diff
@@ -55,28 +50,33 @@ export default class UserListCard extends Component {
...
 
     return (
       <View tabLabel="People" style={styles.container}>
-        <SearchInput onChange={this.handleFilterChange} />
-        <UserList
-          ownEmail={ownEmail}
-          users={users}
-          presence={presence}
-          filter={filter}
+        <Button
```
Comment (neerajwahi) [inline src/conversations/ConversationsCard.js]: I don't like have a button called 'Search' here.

If we want to keep private conversations in the right sidebar, we probably want a tab group to switch between private conversations and the list of all users.

Alternatively, we could put the list of private conversations in the left sidebar (like the web app).

@timabbott @borisyankov thoughts?

File: src/conversations/UserGroup.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  row: {
+    height: 44,
+    flexDirection: 'row',
+    alignItems: 'center',
+    padding: 8,
+  },
+  selectedRow: {
+    backgroundColor: BRAND_COLOR,
+  },
+  touchTarget: {
```
Comment (neerajwahi) [inline src/conversations/UserGroup.js]: Nit: can remove this

File: src/conversations/UserGroup.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    onPress: () => void,
+  }
+
+  handlePress = () =>
+    this.props.onPress(this.props.email);
+
+  render() {
+    const { allNames, isSelected } = this.props;
+
+    return (
```
Comment (neerajwahi) [inline src/conversations/UserGroup.js]: The formatting seems a bit messed up when I look in the simulator. If you have a conversation with many users, the list of users overflows into bordering cells.

File: src/main/MainScreen.js [hunk 8]
```diff
@@ -2,20 +2,23 @@ import React from 'react';
...
   componentDidMount() {
-    const { auth, fetchEvents, fetchUsersAndStatus, fetchMessages } = this.props;
+    const { auth, fetchEvents, fetchUsersAndStatus,
+      backgroundFetchMessages, fetchMessages } = this.props;
 
     fetchEvents(auth);
     fetchUsersAndStatus(auth);
-    fetchMessages(auth, Number.MAX_SAFE_INTEGER, 20, 0, [], [true, true]);
+    fetchMessages(auth, Number.MAX_SAFE_INTEGER, 20, 0, homeNarrow(), [true, true]);
+    backgroundFetchMessages(auth, Number.MAX_SAFE_INTEGER, 100, 0, specialNarrow('private'));
```
Comment (neerajwahi) [inline src/main/MainScreen.js]: @timabbott how hard would it be to add an API endpoint for recent private conversations? Seems cleaner than fetching private messages for that information.

File: src/message-list/messagesActions.js [hunk 9]
```diff
@@ -22,7 +22,7 @@ export const fetchMessages = (
   caughtUp,
 ) =>
   async (dispatch) => {
-    dispatch({ type: MESSAGE_FETCH_START, narrow, fetching, caughtUp });
+    // dispatch({ type: MESSAGE_FETCH_START, narrow });
```
Comment (neerajwahi) [inline src/message-list/messagesActions.js]: Nit: remove commented out code

File: src/title/TitleGroup.js [hunk 10]
```diff
@@ -18,18 +19,18 @@ const styles = StyleSheet.create({
...
-    const { narrow, users } = this.props;
+    const { realm, narrow, users } = this.props;
     const recipientEmails = narrow[0].operand.split(',');
     const recipients = recipientEmails.map(r => users.find(x => x.email === r));
 
     return (
       <View style={styles.wrapper}>
-        {recipients.map(x =>
-          <View key={x.email} style={styles.avatar}>
+        {recipients.map(user =>
```
Comment (neerajwahi) [inline src/title/TitleGroup.js]: I really like renaming `x` to `user`. Much easier to read!

File: src/users/UserList.js [hunk 11]
```diff
@@ -46,13 +42,13 @@ export default class UserList extends Component {
         style={styles.container}
         dataSource={dataSource}
         pageSize={12}
-        renderRow={(x =>
+        renderRow={(user =>
```
Comment (neerajwahi) [inline src/users/UserList.js]: `x` -> `user` 👍

File: src/conversations/UserGroup.js (hunk 7 above)
Comment (borisyankov) [inline src/conversations/UserGroup.js]: Hmm, how many users do you had a conversation with? Do you mean icon or text on the right?

Comment (neerajwahi) [inline src/conversations/UserGroup.js]: I think about 4 (with fairly long usernames).

I mean the text. If it's long enough it wraps and then overflows into neighboring cells.

Comment (borisyankov) [inline src/conversations/UserGroup.js]: You got some long named people. Reproduced and fixed by setting numberOfLines to 2.

File: src/chat/chatSelectors.js (hunk 1 above)
Comment (borisyankov) [inline src/chat/chatSelectors.js]: Reduce returns a single value, which can be an array. In fact `map` and `filter` can both be implemented via `reduce` and that is a valid use.

Still I went with a version similar to your 2nd suggestion as it is simpler.

Comment (borisyankov) [inline src/chat/chatSelectors.js]: Also explicitly including a conversation with self in the list.

**Pull request merged**